src/ai/backend/agent/agent.py

```diff
@@ -202,6 +204,19 @@ def update_additional_gids(environ: MutableMapping[str, str], gids: Iterable[int
     environ["ADDITIONAL_GIDS"] = ",".join(map(str, additional_gids))
 
 
+class ImageCanonicalDigestDict(UserDict):
+    def __setitem__(self, image_canonical: str, image_digest: str) -> None:
+        if not isinstance(image_canonical, str) or not isinstance(image_digest, str):
+            raise TypeError("image_canonical, image_digest must be str")
+        super().__setitem__(image_canonical, image_digest)
+
+
+@dataclass
+class ScanImagesResult:
+    scanned_images: ImageCanonicalDigestDict
+    removed_images: ImageCanonicalDigestDict
```

Comment:
Is there a reason you made it a `UserDict`? I think you could have just used a dict for this part.
My current thought is to create a dataclass with Image canonical and image digest as fields and save it as a set.

Reply:
@dataclass
class ImageMeta
  canonical: str
  digest: str
`dict[str, ImageMeta]`